### PR TITLE
Add Azure file system block

### PR DESCRIPTION
**Pull request by anna-geller**
### Summary
- Adds Azure file system block
- Adds file system blocks docs
- Adds info about file system block types to CLI

### Steps Taken to QA Changes
- created block
- created deployment
- recreated deployment to ensure we overwrite existing (potentially modified) files during a build step

This pull request is:
- [x] A new feature implementation
- [x] Documentation changes + addition

### CODE REVIEW

File: docs/concepts/filesystems.md [hunk 1]
```diff
@@ -67,30 +75,111 @@ How can we use `RemoteFileSystem` to store our flow code?
...
+## Azure
+
+The `Azure` file system block enables interaction with Azure Datake and Azure Blob Storage. Under the hood, `Azure` uses [`adlfs`](https://github.com/fsspec/adlfs).
+
+`Azure` properties include:
+
+| Property | Description                                                                                                                  |
+| --- |------------------------------------------------------------------------------------------------------------------------------|
+| basepath | String path to the location of files on the remote filesystem. Access to files outside of the base path will not be allowed. |
+| azure_storage_connection_string | Azure storage connection string                                                                                              |
```
Comment (tpdorsey) [inline docs/concepts/filesystems.md:161]:
```suggestion
| azure_storage_connection_string | Azure storage connection string. |
```

File: docs/concepts/filesystems.md [hunk 2]
```diff
@@ -67,30 +75,111 @@ How can we use `RemoteFileSystem` to store our flow code?
...
+| azure_storage_account_name | Azure storage account name                                                                                                   |
```
Comment (tpdorsey) [inline docs/concepts/filesystems.md:162]:
```suggestion
| azure_storage_account_name | Azure storage account name. |
```

File: docs/concepts/filesystems.md [hunk 3]
```diff
@@ -67,30 +75,111 @@ How can we use `RemoteFileSystem` to store our flow code?
...
+| azure_storage_account_key | Azure storage account key                                                                                                    |
```
Comment (tpdorsey) [inline docs/concepts/filesystems.md:163]:
```suggestion
| azure_storage_account_key | Azure storage account key. |
```

File: src/prefect/filesystems.py [hunk 4]
```diff
@@ -434,3 +434,94 @@ async def read_path(self, path: str) -> bytes:
 
     async def write_path(self, path: str, content: bytes) -> str:
         return await self.filesystem.write_path(path=path, content=content)
+
+
+class Azure(ReadableFileSystem, WritableFileSystem):
+    """
+    Store data as a file on Azure Datake and Azure Blob Storage.
```
Comment (tpdorsey) [inline src/prefect/filesystems.py:441]: Datalake?

**Review by tpdorsey — APPROVED** [review]

File: docs/concepts/filesystems.md [hunk 5]
```diff
@@ -10,16 +10,19 @@ tags:
 
 # File systems
 
-A file system is an object which allows you to read and write data from paths. Prefect provides two built-in file system types that cover a wide range of use cases. 
+A file system block is an object which allows you to read and write data from paths. Prefect provides two built-in file system types that cover a wide range of use cases.
```
Comment (cicdw) [inline docs/concepts/filesystems.md:13]:
```suggestion
A file system block is an object which allows you to read and write data from paths. Prefect provides multiple built-in file system types that cover a wide range of use cases. 
```

File: docs/concepts/filesystems.md [hunk 6]
```diff
@@ -67,30 +75,111 @@ How can we use `RemoteFileSystem` to store our flow code?
...
+        "secret": "MINIO_ROOT_PASSWORD",
+        "client_kwargs": {"endpoint_url": "http://localhost:9000"},
+    },
 )
+minio_block.save("minio")
 ```
 
-Now let's look at how we can use `RemoteFileSystem` with AWS S3:
+
+## S3
```
Comment (cicdw) [inline docs/concepts/filesystems.md:92]: for the S3, GCS, and Azure blocks could we note somewhere that if credentials are not explicitly set, they will be inferred from the "standard" environment variables for each file system?

**Review by cicdw — COMMENTED** [review]
Very minor doc tweaks but otherwise this LGTM - thank you!!

File: docs/concepts/filesystems.md (hunk 6 above)
Comment (anna-geller) [reply docs/concepts/filesystems.md:92]: both good calls, thanks a lot - will add soon

Comment (anna-geller) [reply docs/concepts/filesystems.md:92]: added a section now

File: docs/concepts/filesystems.md [hunk 7]
```diff
@@ -67,30 +75,116 @@ How can we use `RemoteFileSystem` to store our flow code?
...
+```bash
+prefect deployment build path/to/flow.py:flow_name --name deployment_name --tag dev -sb az/dev
+```
+
+You need to install `adlfs` to use it.
+
+
+## Handling credentials for cloud object storage services
+
+If you leverage `S3`, `GCS`, or `Azure` storage blocks, and you don't explicitly configure credentials on the respective storage block, those credentials will be inferred from the environment. Make sure to set those either on the block or as environment variables, configuration files, or IAM roles within the build environment from which you create your deployments.
```
Comment (cicdw) [inline docs/concepts/filesystems.md:185]:
```suggestion
If you leverage `S3`, `GCS`, or `Azure` storage blocks, and you don't explicitly configure credentials on the respective storage block, those credentials will be inferred from the environment. Make sure to set those either explicitly on the block or as environment variables, configuration files, or IAM roles within both the build and runtime environment for your deployments.
```

**Pull request merged**